### PR TITLE
fix: rule DDLRecommendTableColumnCharsetSame mistriggered 

### PR DESCRIPTION
**Pull request by winfredLIN**
#2043 
## Problem
when the table creation statement declares the column character set and table character set (both are the same), trigger the rule
## Solution
Support checking column character sets in table creation statements
## Support
Add support for the following scenarios
### If the DDL statement is a table creation statement:

1. If both columns and tables declare a character set. The character set of the column is different from the character set of the table, triggering the rule.
2. If the column declares a character set, the table does not declare a character set and only declares sorting rules. The sorting rule corresponds to a character set that is different from the column character set, triggering the rule.
3. If the column declares a character set, the table does not declare a character set and sorting rule. The column character set and the database default character set are different, triggering rules.

### If the DDL statement is a modify table statement

1. If only the column character set is modified. The character set is different from the original table, triggering rules.
2. If you first modify the character set of the table, then modify the character set of the column. The new column character set is different from the new table character set, triggering rules.
3. If you first modify the character set of the column, then modify the character set of the table. The character sets listed in the table are unified as new character sets and do not trigger rules.

### CODE REVIEW

File: sqle/driver/mysql/rule/rule.go [hunk 1]
```diff
@@ -5796,19 +5796,184 @@ func checkColumnQuantity(input *RuleHandlerInput) error {
...
-			if col.Tp.Charset != "" {
+		// 获取建表语句中指定了字符集或排序的列
+		columnWithCharset := getColumnWithCharset(stmt, input)
+		if len(columnWithCharset) == 0 {
+			return nil
+		}
+		// 获取建表语句中的字符集
+		charset := getCharsetFromCreateTableStmt(input.Ctx, stmt)
+		if charset.StrValue == "" {
+			// 未能获取字符集 无法比较 返回
```
Comment (ColdWaterLW) [inline sqle/driver/mysql/rule/rule.go:5842]: 打日志

File: sqle/driver/mysql/rule/rule.go [hunk 2]
```diff
@@ -5796,19 +5796,184 @@ func checkColumnQuantity(input *RuleHandlerInput) error {
...
+			return nil
```
[the remark this answers is not in the document]
Comment (winfredLIN) [reply sqle/driver/mysql/rule/rule.go:5843]: done

File: sqle/driver/mysql/rule/rule.go [hunk 3]
```diff
@@ -5796,19 +5796,184 @@ func checkColumnQuantity(input *RuleHandlerInput) error {
...
 		}
-		return nil
+	case *ast.AlterTableStmt:
+		var columnWithCharset []*ast.ColumnDef
+		var newCharset *ast.TableOption
+		var useConvert bool
+		// 若未更改表的字符集，则获取原表字符集作为表字符集
+		originTable, exist, err := input.Ctx.GetCreateTableStmt(stmt.Table)
+		if err != nil || !exist {
+			return nil
```
[the remark this answers is not in the document]
Comment (winfredLIN) [reply sqle/driver/mysql/rule/rule.go:5858]: done

File: sqle/driver/mysql/rule/rule.go [hunk 4]
```diff
@@ -5796,19 +5796,184 @@ func checkColumnQuantity(input *RuleHandlerInput) error {
...
 				break
 			}
 		}
-		return nil
+	case *ast.AlterTableStmt:
+		var columnWithCharset []*ast.ColumnDef
+		var newCharset *ast.TableOption
+		var useConvert bool
+		// 若未更改表的字符集，则获取原表字符集作为表字符集
+		originTable, exist, err := input.Ctx.GetCreateTableStmt(stmt.Table)
```
[the remark this answers is not in the document]
Comment (winfredLIN) [reply sqle/driver/mysql/rule/rule.go:5856]: done

File: sqle/driver/mysql/rule/rule.go [hunk 5]
```diff
@@ -5796,19 +5796,184 @@ func checkColumnQuantity(input *RuleHandlerInput) error {
...
+						SELECT column_name, character_set_name
+						FROM information_schema.columns
+						WHERE table_name = 'your_table_name';
+					2. 根据SQL语句修改列的字符集到目标字符集
+					3. 判断最终表字符集和最终列字符集是否一致
+				*/
+				return nil
+			}
+		}
+		if newCharset == nil {
```
[the remark this answers is not in the document]
Comment (winfredLIN) [reply sqle/driver/mysql/rule/rule.go:5906]: done

File: sqle/driver/mysql/rule/rule.go [hunk 6]
```diff
@@ -5796,19 +5796,184 @@ func checkColumnQuantity(input *RuleHandlerInput) error {
...
+			if useConvert {
+				for _, column := range columnWithCharset {
+					if column.Tp.Charset != newCharset.StrValue {
+						addResult(input.Res, input.Rule, input.Rule.Name)
+						break
+					}
+				}
+				return nil
+			}
+			if !useConvert {
```
[the remark this answers is not in the document]
Comment (winfredLIN) [reply sqle/driver/mysql/rule/rule.go:5890]: done

File: sqle/driver/mysql/rule/rule.go [hunk 7]
```diff
@@ -5796,19 +5796,184 @@ func checkColumnQuantity(input *RuleHandlerInput) error {
...
+		if newCharset == nil {
+			if len(columnWithCharset) == 0 {
+				// 没有指定字符集的列时，列的字符集被设定为表字符集
+				return nil
+			}
+			// 若没有修改表字符集
+			charset := getCharsetFromCreateTableStmt(input.Ctx, originTable)
+			if charset.StrValue == "" {
+				// 未能获取字符集 无法比较 返回
+				return nil
```
[the remark this answers is not in the document]
Comment (winfredLIN) [reply sqle/driver/mysql/rule/rule.go:5915]: done

**Review by ColdWaterLW — APPROVED** [review]

**Pull request merged**